samcli/lib/build/build_graph.py

```diff
@@ -0,0 +1,188 @@
+    """
+    Converts given build_definition into toml table representation
+
+    :param build_definition: BuildDefinition
+    :return: toml table of BuildDefinition
+    """
+    toml_table = tomlkit.table()
+    toml_table[CODE_URI_FIELD] = build_definition.codeuri
+    toml_table[RUNTIME_FIELD] = build_definition.runtime
+    toml_table[FUNCTIONS_FIELD] = \
+        list(map(lambda f: f.functionname, build_definition.functions))
+
+    if build_definition.metadata:
+        toml_table[METADATA_FIELD] = build_definition.metadata
+
+    return toml_table
+
+
+def _toml_table_to_build_definition(uuid, toml_table):
+    """
+    Converts given toml table into BuildDefinition instance
+
+    :param uuid: key of the toml_table instance
+    :param toml_table: build definition as toml table
+    :return: BuildDefinition of given toml table
+    """
+    build_definition = BuildDefinition(toml_table[RUNTIME_FIELD],
+                                       toml_table[CODE_URI_FIELD],
+                                       dict(toml_table.get(METADATA_FIELD, {})))
+    build_definition.uuid = uuid
+    return build_definition
+
+
+class BuildGraph:
+    """
+    Contains list of build definitions, with ability to read and write them into build.toml file
+    """
+
+    # global table build definitions key
+    BUILD_DEFINITIONS = "build_definitions"
+
+    def __init__(self, build_dir):
+        # put build.toml file inside .aws-sam folder
+        self._filepath = Path(build_dir).parent.joinpath(DEFAULT_BUILD_GRAPH_FILE_NAME)
+        self._build_definitions = None
+        self._read()
+
+    def get_build_definitions(self):
+        return tuple(self._build_definitions)
+
+    def put_build_definition(self, build_definition, function):
+        """
+        Puts the newly read build definition into existing build graph.
+        If graph already contains a build definition which is same as the newly passed one, then it will add
+        the function to the existing one, discarding the new one
+
+        If graph doesn't contain such unique build definition, it will be added to the current build graph
+
+        :param build_definition: build definition which is newly read from template.yaml file
+        :param function: function details for this build definition
+        """
+        if build_definition in self._build_definitions:
+            previous_build_definition = self._build_definitions[self._build_definitions.index(build_definition)]
+            LOG.debug("Same build definition found, adding function (Previous: %s, Current: %s, Function: %s)",
+                      previous_build_definition, build_definition, function)
+            previous_build_definition.functions.append(function)
+        else:
+            LOG.debug("Unique build definition found, adding as new (Build Definition: %s, Function: %s)",
+                      build_definition, function)
+            build_definition.functions.append(function)
```

Comment:
We have add_function() already I think. Here and L124.

Reply:
Good catch will update it